waku/common/enr.nim

```diff
@@ -0,0 +1,107 @@
+## An extension wrapper around nim-eth's ENR module
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+
+import
+  std/options,
+  stew/results,
+  stew/shims/net,
+  eth/keys as eth_keys,
+  eth/p2p/discoveryv5/enr,
+  libp2p/crypto/crypto as libp2p_crypto
+
+export enr
+
+
+## Builder
+
+type EnrBuilder* = object
+    seqNumber: uint64
+    privateKey: eth_keys.PrivateKey
+    fields: seq[FieldPair]
+
+
+proc init*(T: type EnrBuilder, key: eth_keys.PrivateKey, seqNum: uint64 = 1): EnrBuilder =
+  EnrBuilder(
+    seqNumber: seqNum,
+    privateKey: key,
+    fields: newSeq[FieldPair]()
+  )
+
+proc init*(T: type EnrBuilder, key: libp2p_crypto.PrivateKey, seqNum: uint64 = 1): EnrBuilder =
+  # TODO: Inconvenient runtime assertion. Move this assertion to compile time
+  if key.scheme != PKScheme.Secp256k1:
+    raise newException(Defect, "invalid private key scheme")
+
+  let
+    bytes = key.getRawBytes().expect("Private key is valid")
+    privateKey = eth_keys.PrivateKey.fromRaw(bytes).expect("Raw private key is of valid length")
+
+  EnrBuilder.init(key=privateKey, seqNum=seqNum)
+
+proc addFieldPair*(builder: var EnrBuilder, pair: FieldPair) =
```

Comment:
> should we make this private? seems too low level.

I see no harm in it.

I need to double check Nim module visibility rules. If this proc can be accessed from an external module (e.g., `waku_enr`) without being publicly exposed, we can think of making it private. Otherwise, it will have to remain public.

> I mean, we dont want people to build it like `.withIpAddressAndPorts().addFieldPair()`

Why not?

> like `.withIpAddressAndPorts().addFieldPair()`

Note that the `addFieldPair()` proc does not return the builder. What you wrote is not a valid builder extension usage.

This is deliberate. 

Parameters validation, in the case of being necessary, should happen within the `addXXX` extension method and return an `EnrResult[void]` explaining the validation failure cause.

Reply:
> I need to double check Nim module visibility rules. If this proc can be accessed from an external module (e.g., `waku_enr`) without being publicly exposed, we can think of making it private. Otherwise, it will have to remain public.

I can confirm that hiding this method breaks the implementation. It has to be publicly accessible.